### PR TITLE
refactor(headless,atomic): refactor template-provider

### PR DESCRIPTION
**Pull request by olamothe**
First step to refactor atomic-common-result-list. Result list in Atomic relies on some Headless code to organize and evaluate condition against results (or, product in the case of commerce).

* Genericize Headless `result-templates-manager` into a `template-manager`.
* Old `result-templates-manager` is should have no breaking changes, and use the generic `template-manager`.
* Create new Headless `product-templates-manager`, and export it from `commerce.index`
* Genericize Atomic `result-template-provider` into a `template-provider`
* Old Atomic `result-template-provider` should have no breaking changes and use the generic `template-provider`.
* Create new Atomic `product-template-provider`.

https://coveord.atlassian.net/browse/KIT-3033

### CODE REVIEW

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+export type ProductTemplate<Content = unknown> = Template<
+  ProductRecommendation,
+  Content
+>;
+export type ProductTemplateCondition = TemplateCondition<ProductRecommendation>;
+
+export interface ProductTemplatesManager<Content = unknown> {
+  /**
+   * Registers any number of result templates in the manager.
+   * @param templates (...ResultTemplate<Content>) A list of templates to register.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/product-templates/product-templates-manager.ts:17]:
```suggestion
   * @param templates (...Template<Product, Content>) A list of templates to register.
```

(I'm preemptively suggesting Product as the documented type since this is about to change anyways).

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+export type ProductTemplate<Content = unknown> = Template<
+  ProductRecommendation,
+  Content
+>;
+export type ProductTemplateCondition = TemplateCondition<ProductRecommendation>;
+
+export interface ProductTemplatesManager<Content = unknown> {
+  /**
+   * Registers any number of result templates in the manager.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/product-templates/product-templates-manager.ts:16]:
```suggestion
   * Registers any number of product templates in the manager.
```

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+   * @param templates (...ResultTemplate<Content>) A list of templates to register.
+   */
+  registerTemplates(
+    ...newTemplates: Template<ProductRecommendation, Content>[]
+  ): void;
+  /**
+   * Selects the highest priority template for which the given result satisfies all conditions.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/product-templates/product-templates-manager.ts:23]:
```suggestion
   * Selects the highest priority template for which the given product satisfies all conditions.
```

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+   * In the case where satisfied templates have equal priority, the template that was registered first is returned.
+   * @param result (Result) The result for which to select a template.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/product-templates/product-templates-manager.ts:25]:
```suggestion
   * @param product (Product) The product for which to select a template.
```

(once again, preemptive documented type name suggestion).

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+   * @returns (Content) The selected template's content, or null if no template's conditions are satisfied.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/product-templates/product-templates-manager.ts:26]:
```suggestion
   * @returns (Content) The content of the selected template, or null if no template can be selected for the given product.
```

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+   */
+  selectTemplate(product: ProductRecommendation): Content | null;
+}
+
+/**
+ * A manager in which product templates can be registered and selected based on a list of conditions and priority.
+ * @param engine (HeadlessEngine) The `HeadlessEngine` instance of your application.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/product-templates/product-templates-manager.ts:33]:
```suggestion
```

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ * @returns (ResultTemplatesManager<Content, State>) A new result templates manager.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/product-templates/product-templates-manager.ts:34]:
```suggestion
 * @returns (ProductTemplateManager<Content>) A new product template manager.
```

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  TemplateCondition,
+} from '../../templates/templates-manager';
+
+export type ProductTemplate<Content = unknown> = Template<
+  ProductRecommendation,
+  Content
+>;
+export type ProductTemplateCondition = TemplateCondition<ProductRecommendation>;
+
+export interface ProductTemplatesManager<Content = unknown> {
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/product-templates/product-templates-manager.ts:14]:
```suggestion
export interface ProductTemplateManager<Content = unknown> {
```

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+   */
+  selectTemplate(product: ProductRecommendation): Content | null;
+}
+
+/**
+ * A manager in which product templates can be registered and selected based on a list of conditions and priority.
+ * @param engine (HeadlessEngine) The `HeadlessEngine` instance of your application.
+ * @returns (ResultTemplatesManager<Content, State>) A new result templates manager.
+ */
+export function buildProductTemplatesManager<
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/product-templates/product-templates-manager.ts:36]:
```suggestion
export function buildProductTemplateManager<
```

File: packages/headless/src/features/templates/templates-manager.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+export interface TemplatesManager<ItemType, TemplateContent = unknown> {
+  registerTemplates(
+    ...newTemplates: Template<ItemType, TemplateContent>[]
+  ): void;
+
+  selectTemplate(item: ItemType): TemplateContent | null;
+}
+
+export function buildTemplatesManager<
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/templates/templates-manager.ts:50]:
```suggestion
export function buildTemplateManager<
```

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+   * Selects the highest priority template for which the given result satisfies all conditions.
+   * In the case where satisfied templates have equal priority, the template that was registered first is returned.
+   * @param result (Result) The result for which to select a template.
+   * @returns (Content) The selected template's content, or null if no template's conditions are satisfied.
+   */
+  selectTemplate(product: ProductRecommendation): Content | null;
+}
+
+/**
+ * A manager in which product templates can be registered and selected based on a list of conditions and priority.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/product-templates/product-templates-manager.ts:32]:
```suggestion
 * A manager in which product templates can be registered and selected based on a list of conditions and a priority index.
```

File: packages/headless/src/features/templates/templates-manager.ts [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+  /**
+   * The stored content of the template.
+   */
+  content: Content;
+  /**
+   * A list of conditions that must be fulfilled for this template to be selected.
+   */
+  conditions: TemplateCondition<ItemType>[];
+  /**
+   * A value which the manager will fallback to when multiple templates' conditions are fulfilled.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/templates/templates-manager.ts:22]:
```suggestion
   * A value which the manager will use to determine which template to select when an item satisfies the conditions of more than one template.
```

File: packages/headless/src/features/templates/templates-manager.ts [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+   * Templates with higher priority values will be selected over others. The minimum value is `0`.
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/templates/templates-manager.ts:23]:
```suggestion
   * Templates with higher priority values will be selected over others. The minimum and default value is `0`.
```

File: packages/headless/src/features/templates/templates-manager.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+  content: new Value({required: true}),
+  conditions: new Value({required: true}),
+  priority: new NumberValue({required: false, default: 0, min: 0}),
+  fields: new ArrayValue({
+    required: false,
+    each: requiredNonEmptyString,
+  }),
+});
+
+export interface TemplatesManager<ItemType, TemplateContent = unknown> {
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/templates/templates-manager.ts:42]:
```suggestion
export interface TemplateManager<ItemType, TemplateContent = unknown> {
```

File: packages/headless/src/features/templates/templates-manager.ts [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+  ) => {
+    templates.forEach((template) => {
+      templateSchema.validate(template);
+      const areConditionsValid = template.conditions.every(
+        (condition) => condition instanceof Function
+      );
+
+      if (!areConditionsValid) {
+        throw new SchemaValidationError(
+          'Each product template conditions should be a function that takes a result as an argument and returns a boolean'
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/templates/templates-manager.ts:66]:
```suggestion
          'Each template condition should be a function that takes a Result or Product as an argument and returns a boolean'
```

File: packages/headless/src/features/templates/templates-manager.ts [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+  };
+
+  return {
+    registerTemplates(...newTemplates: Template<ItemType, TemplateContent>[]) {
+      validateTemplates(newTemplates);
+
+      newTemplates.forEach((template) => {
+        const templatesWithDefault = {
+          ...template,
+          priority: template.priority || 0,
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/templates/templates-manager.ts:79]: Is the `|| 0` necessary given that there's a default value in the templateSchema?

**Review by fbeaudoincoveo — CHANGES_REQUESTED** [review]

File: packages/headless/src/features/templates/templates-manager.ts (hunk 14 above)
Comment (olamothe) [reply packages/headless/src/features/templates/templates-manager.ts:42]: It is managing multiple templates though...
And the old interface was(and still is) `ResultTemplatesManager` so I would not change this one, so that it fit with what gets exported 🤷

File: packages/headless/src/features/templates/templates-manager.ts (hunk 10 above)
Comment (olamothe) [reply packages/headless/src/features/templates/templates-manager.ts:50]: Same as above.

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts (hunk 8 above)
Comment (olamothe) [reply packages/headless/src/features/commerce/product-templates/product-templates-manager.ts]: Reverting changes due to below explanation

File: packages/headless/src/features/commerce/product-templates/product-templates-manager.ts (hunk 9 above)
Comment (olamothe) [reply packages/headless/src/features/commerce/product-templates/product-templates-manager.ts]: Reverting changes due to below explanation

File: packages/headless/src/features/templates/templates-manager.ts (hunk 16 above)
Comment (olamothe) [reply packages/headless/src/features/templates/templates-manager.ts:75]: "No its not necessary", but it's just old code that was moved around (it's not new). 

I can change it with a cast + use modified returned object from bueno validate I guess...

TLDR: Bueno typing is not advanced enough to do without a cast. Hence this code was explicitly telling Typescript that priority will be defined.

Long explanation:
* `templates` array is `Required<Template<ItemType, TemplateContent>>[]`: All properties must be defined.
*  templateSchema / bueno .validate does not modify original object, returns a copy (IIRC ?).
*  bueno typing is not powerful enough so that if you pass an object/type with something optional, but validate with a `defaultValue` schema, and use the returned object instead of the original, TypeScript won't know automatically that `priority` is now non-optional.

**Review by fbeaudoincoveo — APPROVED** [review]
